atlasdb-service/src/main/java/com/palantir/atlasdb/jackson/TableRowSelectionDeserializer.java

```diff
@@ -42,10 +44,19 @@ public TableRowSelection deserialize(JsonParser jp, DeserializationContext ctxt)
         TableMetadata metadata = metadataCache.getMetadata(tableName);
         Iterable<byte[]> rows = AtlasDeserializers.deserializeRows(metadata.getRowMetadata(), node.get("rows"));
         Iterable<byte[]> columns = AtlasDeserializers.deserializeNamedCols(metadata.getColumns(), node.get("cols"));
+        Collection<byte[]> rowCollection = iterableToCollection(rows);
         if (Iterables.isEmpty(columns)) {
-            return new TableRowSelection(tableName, rows, ColumnSelection.all());
+            return new TableRowSelection(tableName, rowCollection, ColumnSelection.all());
         } else {
-            return new TableRowSelection(tableName, rows, ColumnSelection.create(columns));
+            return new TableRowSelection(tableName, rowCollection, ColumnSelection.create(columns));
         }
     }
+
+    private static <T> Collection<T> iterableToCollection(Iterable<T> iterable) {
+        Collection<T> collection = new ArrayList<T>();
+        for (T item : iterable) {
+            collection.add(item);
+        }
+        return collection;
+    }
```

Comment:
I think we could avoid the full collection copy (or at least do a direct sized copy via ArrayList constructor) by making `AtlasDeserializers.deserializeRows` return `Collection<byte[]>` as `JsonNodeIterable` could determine the size of nodes using something like below.

```
    private static class JsonNodeCollection<T> extends AbstractCollection<T> implements Collection<T> {
        private final JsonNode node;
        private final java.util.function.Function<JsonNode, T> generator;
        private final int size;

        JsonNodeCollection(JsonNode node, java.util.function.Function<JsonNode, T> generator) {
            this.node = node;
            this.generator = generator;
            this.size = (node == null) ? 0 : node.size();
        }

        @Override
        public int size() {
            return size;
        }

        @Override
        public Iterator<T> iterator() {
            if (node == null) {
                return Collections.emptyIterator();
            }
            return new AbstractIterator<T>() {
                private int index = 0;

                @Override
                protected T computeNext() {
                    JsonNode subNode = node.get(index++);
                    if (subNode == null) {
                        return endOfData();
                    }
                    return generator.apply(subNode);
                }
            };
        }

        @Override
        public String toString() {
            return Iterables.toString(this);
        }
    }
```